File: backend/LeetWars.Core/LeetWars.Core.DAL/Entities/Challenge.cs
using LeetWars.Core.DAL.Enums;

namespace LeetWars.Core.DAL.Entities
{
    public class Challenge : AuditEntity<long>
    {
        public string Title { get; set; }
        public string Instructions { get; set; }
        public long LevelId { get; set; }
        public ChallengeLevel? Level { get; set; }

        public ChallengeCategory Category { get; set; }
        public User? Author { get; set; }
        public ICollection<Tag> Tags { get; } = new List<Tag>();

        public ICollection<ChallengeTag> ChallengeTags { get; } = new List<ChallengeTag>();

        public ICollection<ChallengeVersion> Versions { get; } = new List<ChallengeVersion>();
        public ICollection<ChallengeStar> Stars { get; } = new List<ChallengeStar>();

        public Challenge(string title, string instructions)
        {
            Title = title;
            Instructions = instructions;
        }
    }
}

File: frontend/src/app/shared/models/challenge/challenge.ts
import { CategoryType } from '@shared/enums/category-type';

import { IChallengeLevel } from '../challenge-level/challenge-level';
import { IStar } from '../challenge-star/star';
import { IChallengeVersion } from '../challenge-version/challenge-version';
import { AuditEntity } from '../common/audit-entity/audit-entity';
import { ITag } from '../tag/tag';
import { IUser } from '../user/user';

export interface IChallenge extends AuditEntity<number> {
    id: number;
    title: string;
    instructions: string;
    levelId: number;
    level?: IChallengeLevel;
    category: CategoryType;
    author: IUser;
    tags: Array<ITag>;
    versions: Array<IChallengeVersion>;
    stars: Array<IStar>;
}
